### DISCUSSION TITLE
Feature Addition: Set Default Speaker from Filepath or URI in Pipeline

### DISCUSSION THREAD

**Discussion by signalprime** [collabora/WhisperSpeech#128] [Ideas]
I wanted to ask about creating a small PR for adding a function to the `Pipeline` class which allows users to set a default speaker prior to generation.  It could streamlines the process of generating audio with a consistent speaker voice by eliminating the need to repeatedly extract speaker embeddings for each generation task.

I updated [pipeline.py on my fork](https://github.com/signalprime/WhisperSpeech/blob/main/whisperspeech/pipeline.py) and would apply the changes to `/nbs/7. Pipeline.ipynb` if this suggestion is acceptable.  I also updated [pipeline.py](https://github.com/signalprime/WhisperSpeech/blob/main/whisperspeech/pipeline.py) with docstrings and can make a separate PR to present docstrings for the full codebase if it will be helpful.

### Background

Frequently generating audio with the same cloned speaker voice in various applications can lead to inefficiencies due to the repeated extraction of speaker embeddings. This update addresses this issue by allowing a filepath or URI pointing to an audio sample of the desired speaker voice to be set as the default. This change significantly improves performance and user experience by reducing redundant computations.

### Implementation

The method `set_default_speaker` has been introduced, permitting the specification of a local filepath or remote URI that links to an audio file. This file should contain speech from the voice intended to be cloned for text-to-speech generations. Once set, this speaker's embeddings are extracted and used as the default for all subsequent audio generation tasks, thereby optimizing the process.

```
    def set_default_speaker(self, speaker: str) -> None:
        """
        Sets the default speaker for the pipeline.

        This method allows users to specify a default speaker whose embeddings will be used for
        audio generation tasks. It is particularly useful for scenarios where the same speaker's
        voice is used repeatedly, optimizing the process by extracting speaker embeddings only once.

        Args:
            speaker (str): The filename or URI for the speaker whose embeddings are to be extracted 
                          and set as the default.

        Returns:
            None
        """      
        self.default_speaker = self.extract_spk_emb(speaker)
```

### Usage

To set the default speaker, pass a filepath or URI as the argument:
```
# Instantiate the Pipeline
pipe = Pipeline(s2a_ref='collabora/whisperspeech:s2a-q4-base-en+pl.model')

# Set the default speaker for the instance named `pipe`
pipe.set_default_speaker(speaker='https://upload.wikimedia.org/wikipedia/commons/7/75/Winston_Churchill_-_Be_Ye_Men_of_Valour.ogg')

# Generate audio using the new default speaker
audio_tensor = pipe.generate("This is some sample text.")

# Generate additional audio with the custom speaker without reprocessing `atoks` for the custom voice
audio_tensor = pipe.generate("This is some sample text.")
```

**Comment by signalprime** [comment]
@jpc would this be alright with you?

**Comment by jpc** [comment]
Yup, that sounds like a great idea. I you could submit the PR I'll be happy to merge it.
